blockchain/state_transition.go

```diff
@@ -347,11 +358,68 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 
 	rules := st.evm.ChainConfig().Rules(st.evm.Context.BlockNumber)
 
+	// Verify authorization list is not empty.
+	if msg.AuthorizationList() != nil && len(msg.AuthorizationList()) == 0 {
+		return nil, fmt.Errorf("%w: address %v", ErrEmptyAuthList, msg.ValidatedSender())
+	}
+
 	// Execute the preparatory steps for state transition which includes:
 	// - prepare accessList(post-berlin)
 	// - reset transient storage(eip 1153)
 	st.state.Prepare(rules, msg.ValidatedSender(), msg.ValidatedFeePayer(), st.evm.Context.Coinbase, msg.To(), vm.ActivePrecompiles(rules), msg.AccessList())
 
+	// Check authorizations list validity.
+	if msg.AuthorizationList() != nil {
+		for _, auth := range msg.AuthorizationList() {
+			// Verify chain ID is 0 or equal to current chain ID.
+			if auth.ChainID != big.NewInt(0) && st.evm.ChainConfig().ChainID != auth.ChainID {
+				continue
+			}
+			// Limit nonce to 2^64-1 per EIP-2681.
+			if auth.Nonce+1 < auth.Nonce {
+				continue
+			}
+			// Validate signature values and recover authority.
+			authority, err := auth.Authority()
+			if err != nil {
+				continue
+			}
+			// Check the authority account 1) doesn't have code or has exisiting
+			// delegation 2) matches the auth's nonce
+			st.state.AddAddressToAccessList(authority)
+			code := st.state.GetCode(authority)
+			if _, ok := types.ParseDelegation(code); len(code) != 0 && !ok {
+				continue
+			}
+			if have := st.state.GetNonce(authority); have != auth.Nonce {
+				continue
+			}
+			// If the account already exists in state, refund the new account cost
+			// charged in the initrinsic calculation.
+			if exists := st.state.Exist(authority); exists {
+				st.state.AddRefund(params.CallNewAccountGas - params.TxAuthTupleGas)
+			}
+			st.state.IncNonce(authority)
```

Comment:
Please check if `authority` is not account-updated (e.g., authority is not `AccountKeyWeightedMultiSig`).
If we allow such accounts to be set-coded, then the account can be stolen when the original private key is leaked, even if it is no longer used.

Reply:
On a second thought, it is also the case for Ethereum if the original private key is leaked.
The difference is that the users may think that the original key is "deprecated" when updating the account.

However, I still suggest we block set-code tx on account-updated accounts.
For example, [our secretary account](https://kaiascope.com/tx/0xb17fa3a6de3af1048a8f2a7789e3458e9830f12f428f666dcf2459e5ad1c4930?tabId=accountKeyInfo) (which is the owner of the Voting contract) has been account-updated as a multisig account (3-of-5).

- address: 0x815643D2C645e4cCC39a7Bab3BdB735bEc451899
- owners:
  - 0x77ADe79c884306E0CFBB43F1F28e0ed09a117BdD
  - 0x815643D2C645e4cCC39a7Bab3BdB735bEc451899 (same as address)
  - 0x43C5f274f75aad4C3Bf40CF008e60c96F65dAE1e
  - ...

You can see that the second owner has the original private key of the account, so he can subvert the multisig scheme by himself with EIP-7702.

This is just off the top of my head, so it may need further discussion.

Reply:
That's a great opinion.
Since the signer of Tx is EthereumType Tx, only AccountKeyTypeLegacy is allowed, but the authority KeyType was not restricted.
I'll add validation to allow setcode only for AccountKeyTypeLegacy. Thank you!

Reply:
First of all, can legacy transactions be executed by account-updated accounts? I think this is one of legacy transactions. If so, should this feature be limited for only legacy accounts?

Reply:
> The difference is that the users may think that the original key is "deprecated" when updating the account.

Just to confirm, if we update the key in Kaia, we will no longer be able to sign with the original private key, won't we?

Reply:
> Just to confirm, if we update the key in Kaia, we will no longer be able to sign with the original private key, won't we?

If we update the key in Kaia, we will no longer be able to *send transaction* with the original key.

> In this case, I think what is common to both Ethereum and Kaia is that users must be aware of the importance of keeping their private keys even after the code is set.

I think so too. I think we used to tout account-update feature can be used in case of private key theft.
Perhaps from now on, if a private key is stolen, we should recommend to users that they need to migrate to a new account rather than using account-update feature.

Reply:
This seems like a documentation issue, where do you keep track of tasks related to adding things to the Kaia docs?

Reply:
Perhaps we need a KIP document outlining the difference from Ethereum's EIP-7702?
- e.g. KIP-71 vs. EIP-1559 https://kips.kaia.io/KIPs/kip-71
- e.g. KIP-xxx vs. EIP-2028 https://github.com/kaiachain/kips/pull/23

Reply:
Also adding it to the Kaia doc may be necessary.
e.g) https://docs.kaia.io/learn/transactions/ethereum/#txtypeethereumdynamicfee-

Reply:
This seems to be a little off topic from the PR content, so shall we discuss it somewhere else?

Reply:
Sure. I just noticed that this has been resolved by e314ba1